File: BAEKJOON/14620 -  꽃길/꽃길.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits> // INT_MAX 사용을 위한 헤더 포함

using namespace std;

// 방향 벡터 (상, 하, 좌, 우)
int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0, 0};

// 꽃이 차지하는 5칸의 비용을 계산하는 함수
int calculateCost(vector<vector<int>>& board, int x, int y) {
    int cost = board[x][y]; // 중심 칸의 비용
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        cost += board[nx][ny]; // 상하좌우 비용 추가
    }
    return cost;
}

// 꽃이 화단 범위 내에 있는지 확인하는 함수
bool isValid(int x, int y, int N) {
    return x > 0 && x < N - 1 && y > 0 && y < N - 1;
}

// 꽃이 다른 꽃과 겹치는지 확인하는 함수
bool checkOverlap(vector<vector<bool>>& visited, int x, int y) {
    if (visited[x][y]) return false; // 중심 겹침 검사
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (visited[nx][ny]) return false; // 꽃잎 겹침 검사
    }
    return true;
}

// 꽃을 심고 방문 상태를 처리하는 함수
void placeFlower(vector<vector<bool>>& visited, int x, int y, bool place) {
    visited[x][y] = place; // 중심 칸 방문 처리
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        visited[nx][ny] = place; // 상하좌우 방문 처리
    }
}

// 문제 해결 함수
int solve(int N, vector<vector<int>>& board) {
    vector<pair<int, int>> candidates; // 가능한 꽃 중심 좌표 리스트
    int minCost = INT_MAX; // 최소 비용 초기화

    // 가능한 꽃 중심 좌표를 (1, 1)부터 (N-2, N-2)까지 저장
    for (int x = 1; x < N - 1; x++) {
        for (int y = 1; y < N - 1; y++) {
            candidates.push_back({x, y});
        }
    }

    // 세 개의 꽃 중심을 선택하는 모든 조합 탐색
    int C = candidates.size();
    for (int i = 0; i < C; i++) {
        for (int j = i + 1; j < C; j++) {
            for (int k = j + 1; k < C; k++) {
                vector<vector<bool>> visited(N, vector<bool>(N, false)); // 방문 상태 초기화
                bool valid = true;
                int totalCost = 0;

                // 조합으로 선택된 좌표에 꽃을 심기
                for (auto [x, y] : {candidates[i], candidates[j], candidates[k]}) {
                    if (!isValid(x, y, N) || !checkOverlap(visited, x, y)) {
                        valid = false; // 유효하지 않은 경우
                        break;
                    }
                    totalCost += calculateCost(board, x, y); // 비용 계산
                    placeFlower(visited, x, y, true); // 방문 처리
                }

                if (valid) { // 유효한 경우 최소 비용 갱신
                    minCost = min(minCost, totalCost);
                }
            }
        }
    }

    return minCost; // 최소 비용 반환
}

// 메인 함수
int main() {
    int N;
    cin >> N; // 화단 크기 입력
    vector<vector<int>> board(N, vector<int>(N)); // 화단 비용 배열

    // 화단 비용 입력
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> board[i][j];
        }
    }

    // 최소 비용 계산 및 출력
    cout << solve(N, board) << endl;

    return 0;
}

File: BAEKJOON/14620 -  꽃길/꽃길.py
import sys

input = sys.stdin.read
from itertools import combinations

# 방향 벡터 (상, 하, 좌, 우)
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]


# 꽃이 차지하는 5개의 칸에 대한 비용을 계산하는 함수
def calculate_cost(board, x, y, N):
    cost = board[x][y]  # 꽃의 중심 부분 비용
    # 상하좌우의 꽃잎 위치에 대한 비용을 계산
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        cost += board[nx][ny]
    return cost  # 총 비용 반환


# 꽃이 화단을 벗어나는지 여부를 확인하는 함수
def is_valid(x, y, N):
    # 중심 좌표가 1부터 N-2까지만 가능, 이를 벗어나면 False 반환
    if x <= 0 or x >= N - 1 or y <= 0 or y >= N - 1:
        return False
    return True


# 꽃이 다른 꽃과 겹치는지 확인하는 함수
def check_overlap(visited, x, y):
    # 중심 부분이 이미 방문된 경우 겹침
    if visited[x][y]:
        return False
    # 상하좌우 꽃잎이 방문되었는지 확인
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if visited[nx][ny]:
            return False
    return True  # 겹치지 않으면 True 반환


# 꽃을 심고 방문 처리를 수행하는 함수
def place_flower(visited, x, y, place):
    visited[x][y] = place  # 중심 부분 방문 처리
    # 상하좌우 꽃잎 부분도 방문 처리
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        visited[nx][ny] = place


# 문제 해결 함수
def solve(N, board):
    # 가능한 꽃의 중심 좌표 후보 생성 (1, 1)부터 (N-2, N-2)까지
    candidates = [(x, y) for x in range(1, N - 1) for y in range(1, N - 1)]
    min_cost = float("inf")  # 최소 비용을 무한대로 초기화

    # 3개의 꽃을 심을 수 있는 모든 조합 생성
    for comb in combinations(candidates, 3):
        visited = [[False] * N for _ in range(N)]  # 방문 여부 초기화
        total_cost = 0  # 현재 조합에 대한 총 비용
        valid = True  # 현재 조합이 유효한지 여부 확인

        # 조합에 있는 각 꽃의 위치에 대해 반복
        for x, y in comb:
            # 화단 범위를 벗어나거나, 꽃이 겹치는 경우 유효하지 않음
            if not is_valid(x, y, N) or not check_overlap(visited, x, y):
                valid = False
                break  # 더 이상 탐색할 필요 없으므로 중단
            # 꽃을 심을 수 있는 경우 비용을 계산하고 방문 처리
            total_cost += calculate_cost(board, x, y, N)
            place_flower(visited, x, y, True)

        # 유효한 조합에 대해 최소 비용을 갱신
        if valid:
            min_cost = min(min_cost, total_cost)

    return min_cost  # 최소 비용 반환


# 입력을 처리하는 메인 함수
def main():
    data = input().splitlines()  # 표준 입력을 한 줄씩 읽음
    N = int(data[0])  # 첫 번째 줄에서 화단의 크기 N을 읽음
    # 나머지 줄에서 화단의 각 지점별 비용을 2차원 리스트로 변환
    board = [list(map(int, line.split())) for line in data[1:]]

    # 최소 비용을 계산하고 출력
    print(solve(N, board))


main()  # 프로그램 실행
